Add customizable prefix to oban-emitted telemetry events

v0idpwn opened the pull request:
Closes #545.

sorentwo commented on lib/oban/telemetry.ex:
```diff
@@ -305,16 +305,20 @@ defmodule Oban.Telemetry do
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level \\ :info, telemetry_prefix \\ [:oban]) do
```
Using two defaults makes it impossible to set the prefix without setting the level. What if we turn this into a keyword list instead?

```elixir
  def attach_default_logger(level) when is_atom(level) do
    attach_default_logger(level: level)
  end

  def attach_default_logger(opts) when is_list(opts) do
    # ...
```

v0idpwn replied:
That's a good idea! I was afraid of breaking compatibility, but your default solves it :)

sorentwo commented on lib/oban/telemetry.ex:
```diff
@@ -370,17 +374,19 @@ defmodule Oban.Telemetry do
-    |> Map.take([:message, :name])
-    |> log_message("circuit:#{event}", level)
+  def handle_event(event_name, measure, meta, level) do
+    case Enum.slice(event_name, -2..-1) do
```
I'm concerned about the performance of `Enum.slice` when processing a _lot_ of events. It's a bit manual, but we could specify clauses for prefixes with 1-3 values and put a limit in the `config` validation. Something like this

```elixir
[_, :job, event] -> # ...
[_, _, :job, event] -> # ...
[_, _, _, :job, event] -> # ...
```

That may be overkill, and I'm curious to see a benchmark that proves `Enum.slice` is good enough. WDYT?

v0idpwn replied:
I was not sure about which was the best solution there. I'll try to come up with a benchmark

v0idpwn replied:
Pattern matching is around 5-6x faster than relying on Enum.slice. 

I had two options: 
- defining a single clause and using a case
- defining 6 clauses

I went with the second one.

sorentwo commented on test/oban/config_test.exs:
```diff
@@ -81,6 +81,13 @@ defmodule Oban.ConfigTest do
+      refute_valid(telemetry_prefix: :oban)
+      refute_valid(telemetry_prefix: ["oban"])
+
+      assert_valid(telemetry_prefix: [:oban])
```
```suggestion
      assert_valid(telemetry_prefix: [:oban])
      assert_valid(telemetry_prefix: [:my_app, :oban])
```

sorentwo reviewed (commented):
Nicely done! You've covered just about everything. I have one suggestion for telemetry and an open performance question about `Enum.slice`. Otherwise, this is good to go 👍

v0idpwn commented:
I think I addressed all your comments. Let me know if there's something else you'd like to see changed :).

Probably more docs need to be updated as well. I don't know what should be done with the countless references to `[:oban, some, event]` telemetry events there.

v0idpwn commented on test/oban/config_test.exs:
```diff
@@ -81,6 +81,15 @@ defmodule Oban.ConfigTest do
       assert_valid(shutdown_grace_period: 10)
     end
 
+    test ":telemetry_prefix is validated as a list of atoms" do
+      refute_valid(telemetry_prefix: :oban)
+      refute_valid(telemetry_prefix: ["oban"])
+      refute_valid(telemetry_prefix: [:more, :than, :three, :elements])
+
+      assert_valid(telemetry_prefix: [:oban])
+      assert_valid(telemetry_prefix: [:my_app, :oban])
+    end
```
Missed a case: `[]`

sorentwo commented on lib/oban/config.ex:
```diff
@@ -205,9 +205,9 @@ defmodule Oban.Config do
+    unless is_list(prefix) and Enum.all?(prefix, &is_atom/1) and length(prefix) in 1..3 do
       raise ArgumentError,
-            "expected :telemetry_prefix to be a list of atoms with size <= 3, got: #{inspect(prefix)}"
+            "expected :telemetry_prefix to be a non-empty list of atoms with less than 3 items, got: #{inspect(prefix)}"
```
```suggestion
            "expected :telemetry_prefix to be a non-empty list of atoms with up to 3 items, got: #{inspect(prefix)}"
```

sorentwo commented on lib/oban/telemetry.ex:
```diff
@@ -302,19 +302,28 @@ defmodule Oban.Telemetry do
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level_or_opts \\ :info)
```
```suggestion
  def attach_default_logger(level_or_opts \\ [logger_level: :info])
```

sorentwo commented on lib/oban/telemetry.ex:
```diff
@@ -302,19 +302,28 @@ defmodule Oban.Telemetry do
   @doc since: "0.4.0"
   @spec attach_default_logger(Logger.level()) :: :ok | {:error, :already_exists}
-  def attach_default_logger(level \\ :info) do
+  def attach_default_logger(level_or_opts \\ [logger_level: :info])
```
We'll need a different typespec here:

```elixir
  @type attach_option :: {:logger_level, Logger.level()} | {:telemetry_prefix, [atom()]}
```

```elixir
  @spec attach_default_logger(Logger.level() | [attach_option()]) :: :ok | {:error, :already_exists}
```

sorentwo commented:
Wonderful, thank you! 💛
